eos-autoupdater/main.c

```diff
@@ -274,6 +316,153 @@ update_stamp_file (guint64 last_successful_update_secs,
   g_debug ("Wrote stamp file. Next update at %s", next_update);
 }
 
+/**
+ * read_poll_results_file:
+ *
+ * Reads the stored poll results file and returns a new #PollResults
+ * instance.
+ *
+ * Returns: (transfer full) (nullable): a newly allocated #PollResults
+ *  or %NULL if the results file could not be read
+ */
+static PollResults *
+read_poll_results_file (void)
+{
+  const gchar *state_dir = get_state_dir ();
+  g_autofree gchar *results_path = NULL;
+  g_autoptr(GFile) results_file = NULL;
+  g_autoptr(GBytes) results_contents = NULL;
+  g_autoptr(GVariant) results_variant = NULL;
+  g_autoptr(GError) error = NULL;
+  guint64 last_changed_usecs = 0;
+  const gchar *update_refspec = NULL;
+  const gchar *update_id = NULL;
+
+  results_path = g_build_filename (state_dir, POLL_RESULTS_NAME, NULL);
+  results_file = g_file_new_for_path (results_path);
+  results_contents = g_file_load_bytes (results_file, NULL, NULL, &error);
+  if (results_contents == NULL)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND)) {
+        critical (EOS_UPDATER_STAMP_ERROR_MSGID,
+                  "Failed to read autoupdater poll results file: %s",
+                  error->message);
+      }
+
+      return NULL;
+    }
+
+  results_variant = g_variant_new_from_bytes (G_VARIANT_TYPE ("a{sv}"),
+                                              results_contents,
+                                              FALSE);
+  g_variant_lookup (results_variant, "LastChangedUsecs", "t", &last_changed_usecs);
+  g_variant_lookup (results_variant, "UpdateRefspec", "&s", &update_refspec);
+  g_variant_lookup (results_variant, "UpdateID", "&s", &update_id);
+
+  return poll_results_new (last_changed_usecs, update_refspec, update_id);
+}
+
+/**
+ * write_poll_results_file:
+ * @results: the #PollResults to store
+ *
+ * Writes @results to the stored poll results file. The poll results
+ * file is encoded in #GVariant dictionary format. The #PollResults
+ * fields correspond to the following entries:
+ *
+ *   * #PollResults:last_changed_usecs: `LastChangedUsecs` (`t`)
+ *   * #PollResults:update_refspec: `UpdateRefspec` (`s`)
+ *   * #PollResults:update_id: `UpdateID` (`s`)
+ */
+static void
+write_poll_results_file (PollResults *results)
+{
+  const gchar *state_dir = get_state_dir ();
+  g_autofree gchar *results_path = NULL;
+  g_autoptr(GFile) results_file = NULL;
+  g_auto(GVariantBuilder) builder = G_VARIANT_BUILDER_INIT (G_VARIANT_TYPE ("a{sv}"));
+  g_autoptr(GVariant) results_variant = NULL;
+  g_autoptr(GBytes) results_contents = NULL;
+  g_autoptr(GError) error = NULL;
+
+  g_assert (results != NULL);
+
+  g_variant_builder_add (&builder, "{s@v}", "LastChangedUsecs",
+                         g_variant_new_variant (g_variant_new_uint64 (results->last_changed_usecs)));
```

Comment:
Is the `g_variant_new_variant()` call needed here? I thought the `@v` format string takes a value (as a `GVariant*`) and wraps it in a variant wrapper.

i.e. I think (but I could easily be wrong, because this is confusing) that the current code will create the following dict (in text format): `{ 'LastChangedUsecs': <<@t 1234>>, 'UpdateRefspec': <<'some-refspec'>>, 'UpdateID': <<'some-id'>>}`

and what we want is:
`{ 'LastChangedUsecs': <@t 1234>, 'UpdateRefspec': <'some-refspec'>, 'UpdateID': <'some-id'>}`

Reply:
After reading the documentation again and writing a little test program, using `@v` results in the variant being used directly instead of having `g_variant_new` create a new variant automatically. This is how it's used in other parts of eos-updater. I don't know why you wouldn't let the variant get created automatically, though. It's also what the [dictionaries format string](https://docs.gtk.org/glib/gvariant-format-strings.html#dictionaries) documentation tells you to do.

Here's my test program and output:
```
#include <glib.h>

int main(int argc, char *argv[])
{
  GVariantBuilder *a, *b;
  GVariant *a_dict, *b_dict;
  gchar *a_str, *b_str;

  a = g_variant_builder_new (G_VARIANT_TYPE ("a{sv}"));
  g_variant_builder_add (a, "{s@v}", "key",
                         g_variant_new_variant (g_variant_new_string ("value")));
  a_dict = g_variant_builder_end (a);
  a_str = g_variant_print (a_dict, FALSE);

  b = g_variant_builder_new (G_VARIANT_TYPE ("a{sv}"));
  g_variant_builder_add (b, "{sv}", "key",
                         g_variant_new_string ("value"));
  b_dict = g_variant_builder_end (b);
  b_str = g_variant_print (b_dict, FALSE);

  g_print ("a: %s\nb: %s\n", a_str, b_str);

  return 0;
}
$ ./vardict 
a: {'key': <'value'>}
b: {'key': <'value'>}
```